### ISSUE TITLE
mcMMO Config Update API reworks

### ISSUE THREAD

**Issue by electronicboy** [mcMMO-Dev/mcMMO#4120]
In the past, mcMMO has never had what many would consider a proper API, there are a few classes which have been exposed which are designed for interacting with a few aspects of the plugin, e.g. fetching player levels, etc, but there has never been a manner to interact with skills and their outcomes, e.g. xp gains, etc

This issue is somewhat of a discussion point for "would be nice to haves", 2.2. will likely release with a minimal API which will slowly be expanded as needs arise and code is refactored

**Comment by PikaMug** [comment]
I am in opposition to this ticket. While myself and the Bukkit/Spigot community as a whole have a high degree of respect for the developers of mcMMO and the project itself, version 2.2.0 has been hyped for well over a year and has become a manner of fantasy in some circles. At the current rate of development, which I have been following closely, I would not expect to see the minor version increment until late-2021 at the earliest.

This is not something I question lightly. [My own project](https://github.com/PikaMug/Quests) of 6.7k commits sees equally small, sporadic changes, as I am its sole lead developer in my spare time. Having been pestered for the 4.0.0 release of my own resource, I understand that there isn't enough time to please everyone.

However, I submitted mcMMO PR #3793 exactly one year _prior_ to the creation of this ticket, with similar goals, and was told it would not be considered until _after_ the 2.2 release was "playable". Adding a milestone and pin to this issue highlights a frequent problem: using 2.2 as an excuse to brush delicate or time-consuming efforts under the rug. We all have real lives and more important things to do, but promoting unlikely goals like this, especially for a now-premium resource, feels detrimental to the community.

**Comment by nossr50** [comment]
There were many things that halted mcMMO development over the last year, I'm just now recovering and getting to a good place mentally so work is picking up again (I've written about what's been going on in my life before). What is being referred to as 2.2 should be instead referred to as the config update, 2.2 right now is likely to be the release of Tridents/Xbows in the near future.

Things are looking good for me now as my GF has almost made a full recovery and I've been seeing an excellent therapist on a weekly basis for the last few months, I'm hoping things will continue to improve and mcMMO dev will be back on track.

The API for mcMMO is in serious needs of a rewrite, it will happen most likely in the config update or the update following that. I would currently expect that the config update may be released in early 2021 if things go well.

**Comment by PikaMug** [comment]
Glad to hear you both are doing better. My last hearing of this was in mid-January, so please forgive me if I've missed out on recent developments. I've solely been following the commit history and seemingly ought to visit Discord more often.

That said, please consider closing this ticket. It's clear you have more realistic goals to follow than crossbows or APIs. You and your team's aspirations are commendable, but putting something so visibly distant at the forefront of discussion has opened the proverbial door to unrealistic expectations from parts of the fanbase, myself included.

If the 2.2.0 milestone is no longer applicable, I feel that's reason enough, but as a beloved developer of such a fine plugin, no-one wants to see you over-extend yourself. Do let me know if you'd like help sorting this issue tracker!

**Comment by electronicboy** [comment]
This issue was created as means to discussion API additions/considerations, especially as I was working on abstracting out the logic for cross platform support; The idea was that we'd end up with something sane that could be used cross platform, which ofc, means ripping out a lot of the existing API classes, thus, I figured having a place for people to raise stuff up would be nice

Due to my health, I sorta ended up having to pull back, as I write now my head wants to crash into the desk; This has seemingly stalled however for varying reasons, closing this issue would have 0 impact bar closing a discussion point for developers

**Comment by PikaMug** [comment]
Yes, I recall hearing of that around April last year, near the release of Minecraft 1.14.
<details>
  <summary>Flashback</summary>

![abstraction](https://i.imgur.com/AZMHKuh.jpg)
</details>



I'm happy that health has been prioritized - as it should be - but still feel a need to question this workflow. At the time of conception, Sponge was seen as a promising newcomer that was just taking its time updating to MC 1.13, but now a whole two years later (as of yesterday), it's pretty clear that it's stuck on 1.8 - 1.12, which mcMMO Overhaul doesn't even support. All remaining target platforms are so niche that I can't imagine the effort being worth it.

Perhaps I take this a bit personally since most of my users run it alongside mcMMO - again, myself included - but it does seem like this is a discussion point worth closing, or at the very least making it clear that it's a _very_ long term goal, far beyond 2.2.

**Comment by nossr50** [comment]
I don't think closing the ticket does anything meaningful, tickets can be an open discussion on an issue or project goal. As for Sponge, the work to port mcMMO over to Sponge will mostly be done by the Sponge devs themselves as they have volunteered for the job, but they are waiting on abstraction in our code base to do so, which will happen when we port support to 1.8/1.12

**Comment by mrfloris** [comment]
This still going to happen? Because **ALL** The suggestions, options, feature requests, and everything always got the response two plus years ago "all easy and possible after the config update in 2.3", but i haven't seen any new features, skills, suggestion implementations, or big config updates or 2.3 happen. and this hasn't had a response for 12 months now. 

Is the mcMMO project dead by now, or on halt until 2023 or what's the current setup for the roadmap that changed a few times but never got a realistic follow up in code? 

I am just getting really worried about the project and making it modern. as i do every 6 or so months. but years are ticking away and i just want to know if we're investing $1000 into configuring things and setting it up with mcmmo api to integrate things into our setups on some servers, or if i need to invest $1000 into hiring a develop to go through our list of things that we wish mcmmo could have/do and split it off to some proprietary product. 

If real life is in the way and big things aren't possible, i can understand this, but i rather then hear in time that mcmmo overhaul will only get patches updates and maybe an exploit fix, versus the things we're waiting and wishing for.
